File: 0404_sum_of_left_leaves/sum_of_leaves.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def sum_of_left_leaves(self, root: TreeNode) -> int:
    """
    Time: O(n)
    Space: O(n)
    """
    return helper(root, 1)


def helper(node: TreeNode, side: bool) -> int:
    if not node:
        return 0
    if not node.right and not node.left:
        return node.val if side == 0 else 0
    return helper(node.left, 0) + helper(node.right, 1)

File: 0404_sum_of_left_leaves/sum_of_leaves.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Integer}
def sum_of_left_leaves(root)
    # Time: O(n)
    # Space: O(n)
    return dfs(root)
end

def dfs(node, is_left = false)
    return 0 if node.nil?
    if node.left.nil? && node.right.nil?
        return is_left ? node.val : 0
    end
    dfs(node.left, true) + dfs(node.right)
end